bindings/c/testc.c

```diff
@@ -0,0 +1,91 @@
+
+#include "types.h"
+
+ManifoldVec3 warp(float x, float y, float z) {
+  ManifoldVec3 res = {0.5 * x + 30, 0.5 * y + 30, z + 30};
+  return res;
+}
+
+const float radius = 15;
+const float xscale = 3;
+const float yscale = 1;
+const float zscale = 1;
+const float bb = radius * 2;
+float sdf_ellipsoid(float x, float y, float z) {
+  float xs = x / xscale;
+  float ys = y / yscale;
+  float zs = z / zscale;
+  return radius - sqrt(xs * xs + ys * ys + zs * zs);
+}
+
+int main(void) {
+  size_t sz = manifold_manifold_size();
+  ManifoldManifold *s1 = manifold_sphere(malloc(sz), 30., 360);
+  ManifoldManifold *s1_trans = manifold_translate(malloc(sz), s1, 15., 0., 0.);
+  ManifoldManifold *s2 = manifold_rotate(malloc(sz), s1_trans, 0., 0., -90.);
+  ManifoldManifold *diff = manifold_difference(malloc(sz), s1, s2);
+  ManifoldManifold *warped = manifold_warp(malloc(sz), diff, warp);
+  ManifoldManifold *final = manifold_union(malloc(sz), diff, warped);
+
+  manifold_delete_manifold(s1);
+  manifold_delete_manifold(s1_trans);
+  manifold_delete_manifold(s2);
+  manifold_delete_manifold(diff);
+  manifold_delete_manifold(warped);
+
+  ManifoldMesh *mesh = manifold_get_mesh(malloc(manifold_mesh_size()), final);
+  ManifoldExportOptions *options =
+      manifold_export_options(malloc(manifold_export_options_size()));
+  manifold_export_mesh("test.stl", mesh, options);
+
+  // extrude
+  ManifoldVec2 pts[] = {{0, 0}, {0, 30}, {30, 30}, {30, 0}};
+  ManifoldSimplePolygon *sq[] = {manifold_simple_polygon(
+      malloc(manifold_simple_polygon_size()), &pts[0], 4)};
+  ManifoldPolygons *polys =
+      manifold_polygons(malloc(manifold_polygons_size()), sq, 1);
+  ManifoldManifold *pillar =
+      manifold_extrude(malloc(sz), polys, 120, 360, 360, 2, 2);
+  ManifoldMesh *pillar_mesh =
+      manifold_get_mesh(malloc(manifold_mesh_size()), pillar);
+  manifold_export_mesh("pillar.stl", pillar_mesh, options);
+
+  // signed distance function
+  ManifoldBox *bounds =
+      manifold_box(malloc(manifold_box_size()), -bb * xscale, -bb * yscale,
+                   -bb * zscale, bb * xscale, bb * yscale, bb * zscale);
+  ManifoldMesh *sdf_mesh = manifold_level_set(malloc(manifold_mesh_size()),
+                                              sdf_ellipsoid, bounds, 0.5, 0);
+  manifold_export_mesh("sdf_test.stl", sdf_mesh, options);
+
+  // getting dynamically sized vector data out
+  ManifoldMeshRelation *rel =
+      manifold_get_mesh_relation(malloc(manifold_mesh_relation_size()), final);
+  size_t bary_len = manifold_mesh_relation_barycentric_length(rel);
+  ManifoldVec3 *barycentric = manifold_mesh_relation_barycentric(
+      malloc(sizeof(ManifoldVec3) * bary_len), rel);
+  size_t print_until = (bary_len > 10 ? 10 : bary_len);
+  for (int i = 0; i < print_until; ++i) {
+    ManifoldVec3 v = barycentric[i];
+    printf("(%f %f %f)\n", v.x, v.y, v.z);
+  }
+  ManifoldVec3 last = barycentric[bary_len - 1];
+  printf("(%f %f %f)\n", last.x, last.y, last.z);
```

Comment:
Should we make some `expect` calls to show when it fails? Does `GTest` have a C version?

Reply:
I'm not totally sure, I hadn't thought farther about the tests other than what I needed to do along the way to make sure things were working as I added them. :sweat_smile:

Happy to hear suggestions on how the testing should be done, laid out. I'm pretty unfamiliar with how it is usually done in C/C++.

Reply:
GTest is the framework we use for the rest of our C++ tests and it looks like it'll be fine for your C too: https://stackoverflow.com/questions/5335268/is-google-test-ok-for-testing-c-code

It's pretty simple, I don't think you'll have any trouble adapting it from our existing tests.

Reply:
Ok I'll give that a shot soon. How much additional coverage beyond what I have in there do you think I should add? Right now it just has a couple basics, plus some of the trickier parts like the sdf.

Reply:
What you have looks good; again since you've signed up to maintain these C bindings, these tests are mostly to make your own life easier. You're welcome to add more as you see fit.

Reply:
Roger. I'll add some more of what I had in `testc.c` to the new `manifoldc_test.cpp` so I can get rid of that lone executable, then populate further as needed. 👍